lib/Dialect/TritonGPU/Transforms/RemoveLayoutConversions.cpp

```diff
@@ -720,7 +720,7 @@ Operation *LayoutPropagation::rewriteOp(Operation *op) {
 
 bool canBeRemat(Operation *op) {
   if (isa<LoadOp, StoreOp>(op))
-    return !isExpensiveLoadOrStore(op);
+    return op->hasOneUse() || !isExpensiveLoadOrStore(op);
```

Comment:
unfortunately I don't think this will always result in better performance. There are many cases where it is better to load with good coalesced accesses and do a convert rather than loading from a layout with worse memory access pattern.
I think a simple transpose kernel would expose this problem:
%v = load %ptr <32x32>
store %v, %transpose_ptr <32x32>
here the fastest implementation it is to coalesce load/store and to have use a convert in between.

Reply:
Yeah, I just realized a similar case in test_core.py.  Do you think considering the target layout, i.e, the layout that the load will be rematerialized into, is better?

Reply:
It's one way. My guess is that this could probably be solved by improving the logic in Coalesce.cpp, I had looked at it at some point but this can be a very sensitive heuristic.
What does the layout look before and after the transformation in your case?

Reply:
Before the transformation, the load of B in ((Q*K)+B)*V is in blocked layout:

`#blocked = #triton_gpu.blocked<{sizePerThread = [1, 8], threadsPerWarp = [4, 8], warpsPerCTA = [4, 1], order = [1, 0], CTAsPerCGA = [1, 1], CTASplitNum = [1, 1], CTAOrder = [1, 0]}>`

After that the load is rematerizlied into a mma layout:

`#mma = #triton_gpu.nvidia_mma<{versionMajor = 2, versionMinor = 0, warpsPerCTA = [1, 4], instrShape = [16, 8]}>
`
This led up to 15% speedup for our model on H100.

Reply:
interesting, I had experimented with doing loads directly in #mma layout before but it was often worse. So it sounds like this load is in the loop? Can we pipeline the load instead and make it an async_cp? In this case the ptr layout should not matter (since it would always go through shared memory)

Reply:
Yes it is in a loop. And actually it is in a chain of loads, similar to the luk load case that was handled by recent pipeliner work.

I'm trying to understand how pipelining will help here. IIUC, the improvement is mostly from avoiding converting from blocked layout to mma layout through SMEM. Without the transformation, `B` is loaded into registers first, and then dumped to SMEM and reloaded from MEM again but to a mma layout. With pipelining we still need the last step right?

Reply:
Also too much use of SMEM could have other effects, especially on AMDGPU where SMEM isn't as big as on H100?

Reply:
> I'm trying to understand how pipelining will help here. 
if we use async copy and pipeline then the load will always go through shared memory so you can use blocked layout for copy to shared then when loading from shared use mma layout.

> Also too much use of SMEM could have other effects, especially on AMDGPU where SMEM isn't as big as on H100?
right, there is a tradeoff, since it is in the main loop pipelining would probably help, maybe it shouldn't be pipelined as aggressively. Note that convert_layout also use shared memory so technically we could pipeline this load a bit without increasing shared memory usage. However we don't do very fine grain pipelining right now.

It is possible that the best solution is to load directly from mma however having an heuristic to decide that sounds tricky.

Reply:
> It is possible that the best solution is to load directly from mma however having an heuristic to decide that sounds tricky.

On the second thought, I was trying to use axis information to determine if a load with a new encoding can be more expensive than the original load, from vectorization POV. However, the axis information does not work well on a mutating IR. Before I dive deeper, I'd like to hear your thoughts on it. Do you think that's a reasonable path to continue looking into?